### DISCUSSION TITLE
Entitie count's confusion

### DISCUSSION THREAD

**Discussion by zkryakgul** [milvus-io/milvus#8932] [Q&A and General discussion]
Hi everyone,

Im trying to fix duplicate entities in my deployment. And i fairly done it with help @yhmo on my previus question in here: https://github.com/milvus-io/milvus/discussions/8846#discussioncomment-1401543

However, something confused me.

When i build a list of all ids and then get length of this list it returns **44853579**.
```
MILVUS = Milvus(host=SERVER_ADDR, port=SERVER_PORT)
def list_ids():
    status, stats = MILVUS.get_collection_stats(collection_name=COLLECTION_NAME)
    part_default = stats["partitions"][0]
    segments = part_default["segments"]
    all_ids = []
    for seg in segments:
        seg_name = seg["name"]
        status, ids = MILVUS.list_id_in_segment(collection_name=COLLECTION_NAME, segment_name=seg_name)
        all_ids = all_ids + ids
    print(len(all_ids))
```

But if I directly call the 
```MILVUS.count_entities(COLLECTION_NAME)```
 then it returns **44853478**

Why this two returns diffrent numbers? What can cause this diffrance?

**Comment by yhmo** [comment]
Is there any operation between you call the list_ids() and MILVUS.count_entities()?

I did some testing but could not reproduce this problem. I think it could be a bug in certain workflow.

**Comment by zkryakgul** [comment]
There is no other operation between this commands. I ran them one after another. But we performed the delete and reinsert operation for duplicated id entities before get this numbers. After that we called the `compact` function. When it finished we get this numbers.

**Reply by ismailocal** [reply]
@yhmo we got the same problem.

**Reply by xiaofan-luan** [reply]
> @yhmo we got the same problem.

which milvus version is that? better to have a related issue with logs

**Comment by yhmo** [comment]
@zkryakgul @ismailocal 
I have found the reason.
Milvus merge small segments into large segment, once a small segment is merged successfully, milvus will delete its metadata and segment files.
The function list_ids() firstly call get_collection_stats(), then call list_id_in_segment(). Merge action may happens between the two methods. Assume the get_collection_stats() return 3 segments, two of them are small segment, milvus merge/delete the two small segments, then when list_id_in_segment() is called, the two segments are no longer available, the list_id_in_segment() will return error "Segment does not exist". This is the reason why the result of list_ids() is not equal to count_entities().

The following script can reprocduce:
```

import time
import numpy as np
from milvus import Milvus
from milvus.client.types import MetricType

SERVER_ADDR = "127.0.0.1"
SERVER_PORT = '19530'
COLLECTION_NAME = "TEST"
COLLECTION_DIMENSION = 16

MILVUS = Milvus(host=SERVER_ADDR, port=SERVER_PORT)


def count_entities():
    status, cnt = MILVUS.count_entities(collection_name=COLLECTION_NAME)
    print("entity count:", cnt)

def gen_vec_list(nb, seed=np.random.RandomState(1234)):
    xb = seed.rand(nb, COLLECTION_DIMENSION).astype("float32")
    vec_list = xb.tolist()
    return vec_list

def prepare():
    s, has = MILVUS.has_collection(collection_name=COLLECTION_NAME)
    if has:
        MILVUS.drop_collection(collection_name=COLLECTION_NAME)
        time.sleep(1)

    MILVUS.create_collection(param={'collection_name': COLLECTION_NAME,
                                'dimension': COLLECTION_DIMENSION,
                                'index_file_size': 1024,
                                'metric_type': MetricType.L2})

def list_ids():
    status, stats = MILVUS.get_collection_stats(collection_name=COLLECTION_NAME)
    print(stats)
    part_default = stats["partitions"][0]
    segments = part_default["segments"]
    all_ids = []
    for seg in segments:
        seg_name = seg["name"]
        status, ids = MILVUS.list_id_in_segment(collection_name=COLLECTION_NAME, segment_name=seg_name)
        print(status)
        if len(ids) == 0:
            print('the segment', seg_name, 'is no longer exist')
        all_ids = all_ids + ids
    print('len of ids', len(all_ids))

if __name__ == "__main__":
    prepare()

    print("===== insert =====")
    batch = 100000
    vec = gen_vec_list(batch)
    for k in range(50):
        print('insert', k)
        ids = [i + k*batch for i in range(batch)]
        MILVUS.insert(collection_name=COLLECTION_NAME, records=vec, ids=ids)

    MILVUS.flush()
    list_ids()
    count_entities()

```
Run this script several times. Some times the script will print error like this:
```
Status(code=0, message='Success')
Status(code=0, message='Success')
Status(code=0, message='Success')
Status(code=1, message='Segment does not exist')
the segment 1634526179638138000 is no longer exist
len of ids 4700000
entity count: 5000000
```

The MILVUS.count_entities() result is correct.
The list_ids() result could wrong if milvus background thread is merging segments.
